Mathlib/Combinatorics/SimpleGraph/Extremal/Basic.lean

```diff
@@ -0,0 +1,177 @@
+
+variable [Fintype V] [DecidableRel G.Adj]
+
+/-- If `G` is `H`-free, then `G` has at most `extremalNumber (card V) H` edges. -/
+theorem card_edgeFinset_le_extremalNumber (h : H.Free G) :
+    #G.edgeFinset ≤ extremalNumber (card V) H := by
+  rw [extremalNumber_of_fintypeCard_eq rfl]
+  convert @le_sup _ _ _ _ { G | H.Free G } (#·.edgeFinset) G (by simpa using h)
+
+/-- If `G` has more than `extremalNumber (card V) H` edges, then `G` contains a copy of `H`. -/
+theorem IsContained.of_extremalNumber_lt_card_edgeFinset
+    (h : extremalNumber (card V) H < #G.edgeFinset) : H ⊑ G := by
+  contrapose! h
+  exact card_edgeFinset_le_extremalNumber h
+
+/-- `extremalNumber (card V) H` is at most `x` if and only if every `H`-free simple graph `G` has
+at most `x` edges. -/
+theorem extremalNumber_le_iff (H : SimpleGraph W) (x : ℕ) :
+    extremalNumber (card V) H ≤ x ↔
+      ∀ (G : SimpleGraph V) [DecidableRel G.Adj], H.Free G → #G.edgeFinset ≤ x := by
+  simp_rw [extremalNumber_of_fintypeCard_eq rfl, Finset.sup_le_iff, mem_filter, mem_univ, true_and]
+  exact ⟨fun h _ _ h' ↦ by convert h _ h', fun h _ h' ↦ by convert h _ h'⟩
+
+/-- `extremalNumber (card V) H` is greater than `x` if and only if there exists a `H`-free simple
+graph `G` with more than `x` edges. -/
+theorem lt_extremalNumber_iff (H : SimpleGraph W) (x : ℕ) :
+    x < extremalNumber (card V) H ↔
+      ∃ G : SimpleGraph V, ∃ _ : DecidableRel G.Adj, H.Free G ∧ x < #G.edgeFinset := by
+  simp_rw [extremalNumber_of_fintypeCard_eq rfl, Finset.lt_sup_iff, mem_filter, mem_univ, true_and]
+  exact ⟨fun ⟨_, h, h'⟩ ↦ ⟨_, _, h, h'⟩, fun ⟨_, _, h, h'⟩ ↦ ⟨_, h, by convert h'⟩⟩
+
+variable {R : Type*} [LinearOrderedSemiring R] [FloorSemiring R]
+
+@[inherit_doc extremalNumber_le_iff]
+theorem extremalNumber_le_iff_of_nonneg (H : SimpleGraph W) {x : R} (h : 0 ≤ x) :
+    extremalNumber (card V) H ≤ x ↔
+      ∀ (G : SimpleGraph V) [DecidableRel G.Adj], H.Free G → #G.edgeFinset ≤ x := by
+  simp_rw [← Nat.le_floor_iff h]
+  exact extremalNumber_le_iff H ⌊x⌋₊
+
+@[inherit_doc lt_extremalNumber_iff]
+theorem lt_extremalNumber_iff_of_nonneg (H : SimpleGraph W) {x : R} (h : 0 ≤ x) :
+    x < extremalNumber (card V) H ↔
+      ∃ G : SimpleGraph V, ∃ _ : DecidableRel G.Adj, H.Free G ∧ x < #G.edgeFinset := by
+  simp_rw [← Nat.floor_lt h]
+  exact lt_extremalNumber_iff H ⌊x⌋₊
+
+/-- If `H` contains a copy of `H'`, then `extremalNumber n H` is at most `extremalNumber n H`. -/
+theorem IsContained.extremalNumber_le {W' : Type*} {H' : SimpleGraph W'} (h : H' ⊑ H) :
+    extremalNumber n H' ≤ extremalNumber n H := by
+  rw [← Fintype.card_fin n, extremalNumber_le_iff]
+  intro _ _ h'
+  contrapose! h'
+  rw [not_not]
+  exact h.trans (IsContained.of_extremalNumber_lt_card_edgeFinset h')
+
+/-- If `H₁ ≃g H₂`, then `extremalNumber n₁ H₁` equals `extremalNumber n₂ H₂`. -/
+@[congr]
+theorem extremalNumber_congr {n₁ n₂ : ℕ} {W₁ W₂ : Type*} {H₁ : SimpleGraph W₁}
```

Comment:
Can we have the `congr_right` version of this, where n is fixed, for ease of use in rewriting